miden-crypto/src/ies/crypto_box.rs

```diff
@@ -0,0 +1,161 @@
+//! Core cryptographic primitive for Integrated Encryption Scheme (IES).
+//!
+//! This module defines the generic `CryptoBox` abstraction that combines a key agreement scheme
+//! (e.g. K256 ECDH) with an AEAD scheme (e.g. XChaCha20-Poly1305) to provide authenticated
+//! encryption.
+//!
+//! It also defines the `RawSealedMessage` which carries ephemeral keys, nonce, and ciphertext
+//! in raw form.
+
+use alloc::vec::Vec;
+
+use rand::{CryptoRng, RngCore};
+use zeroize::Zeroize;
+
+use super::error::IntegratedEncryptionSchemeError;
+use crate::{
+    Felt,
+    aead::AeadScheme,
+    ecdh::KeyAgreementScheme,
+    utils::{Deserializable, Serializable},
+};
+
+/// A generic CryptoBox primitive parameterized by key agreement and AEAD schemes
+pub(crate) struct CryptoBox<K: KeyAgreementScheme, A: AeadScheme> {
+    _phantom: core::marker::PhantomData<(K, A)>,
+}
+
+/// Internal raw sealed message representation
+#[derive(Debug)]
+pub(crate) struct RawSealedMessage {
+    pub ephemeral_public_key: Vec<u8>,
+    pub ciphertext: Vec<u8>,
+}
+
+impl<K: KeyAgreementScheme, A: AeadScheme> CryptoBox<K, A> {
+    // BYTE-SPECIFIC METHODS
+    // ================================================================================================
+
+    pub(crate) fn seal_with_associated_data<R: CryptoRng + RngCore>(
+        rng: &mut R,
+        recipient_public_key: &K::PublicKey,
+        plaintext: &[u8],
+        associated_data: &[u8],
+    ) -> Result<RawSealedMessage, IntegratedEncryptionSchemeError> {
+        let (ephemeral_private, ephemeral_public) = K::generate_ephemeral_keypair(rng);
+
+        let shared_secret = K::exchange_ephemeral_static(ephemeral_private, recipient_public_key)
+            .map_err(|_| IntegratedEncryptionSchemeError::KeyAgreementFailed)?;
+
+        let encryption_key_bytes =
+            K::extract_key_material(&shared_secret, <A as AeadScheme>::KEY_SIZE)
+                .map_err(|_| IntegratedEncryptionSchemeError::FailedExtractKeyMaterial)?;
+        let mut encryption_key = A::key_from_bytes(&encryption_key_bytes)
+            .map_err(|_| IntegratedEncryptionSchemeError::EncryptionKeyCreationFailed)?;
+
+        let ciphertext = A::encrypt_bytes(&encryption_key, rng, plaintext, associated_data)
+            .map_err(|_| IntegratedEncryptionSchemeError::EncryptionFailed)?;
+
+        encryption_key.zeroize();
```

Comment:
I notice the manual zeroization pattern in the CryptoBox methods where you call `encryption_key.zeroize()` and `decryption_key.zeroize()` immediately after use.

Have you considered if there are any code paths where an early return could bypass the zeroization?
Would using a `Zeroizing<T>` wrapper or a scope-based approach to guarantee cleanup even on errors?

Reply:
Good point! I have now used the wrapper zeroizing struct in order to make sure we drop even on early returns.